### PR TITLE
Thycotic secret server

### PR DESCRIPTION
**Pull request by AndyNikolaev**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 1]
```diff
@@ -0,0 +1,1096 @@
...
+    - default: false
+      description: Sort direction
+      isArray: false
+      name: sortBy[0]_direction
+      required: false
+      secret: false
+    - default: false
+      description: Sort field name
+      isArray: false
+      name: sortBy[0]_name
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:873]: This (and similar) are not valid names for arguments, let's fix them

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 2]
```diff
@@ -0,0 +1,1096 @@
...
+      required: false
+      secret: false
+    - default: false
+      description: Priority index. Sorts with lower values are executed earlier (integer)
+      isArray: false
+      name: sortBy[0]_priority
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:879]: same here

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.py [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+        'thycotic-secret-username-get': secret_username_get_command,
+        'thycotic-secret-get': secret_get_command,
+        'thycotic-secret-search-name': secret_search_name_command,
+        'thycotic-secret-search': secret_search_command,
+        'thycotic-secret-password-update': secret_password_update_command,
+        'thycotic-secret-checkout': secret_checkout_command,
+        'thycotic-secret-checkin': secret_checkin_command,
+        'thycotic-secret-create': secret_create_command,
+        'thycotic-secret-delete': secret_delete_command,
+        'thycotic-secret-rpc-changepassword': secret_rpc_changepassword_command,
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.py:500]: What does this command do?

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 4]
```diff
@@ -0,0 +1,1096 @@
...
+      name: newpassword
+      required: true
+      secret: false
+    deprecated: false
+    description: Remote Password Changing
+    execution: false
+    name: thycotic-secret-rpc-changepassword
+    outputs:
+    - contextPath: Thycotic.Secret.ChangePassword
+      description: Secret summary
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1083]: I don't understand this description, what can be found here?

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 5]
```diff
@@ -0,0 +1,1096 @@
...
+      secret: false
+    - default: false
+      description: GroupOwners(integer)
+      isArray: false
+      name: groupowners
+      required: false
+      secret: false
+    - auto: PREDEFINED
+      default: false
+      description: IsApplicationAccount
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:959]: Many of the arguments' descriptions need to be replaced.
Some examples can be found below (I'll add them as notes), but please review all of them and replace them with meaningful, detailed descriptions

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 6]
```diff
@@ -0,0 +1,1096 @@
...
+      isArray: false
+      name: isapplicationaccount
+      predefined:
+      - 'true'
+      - 'false'
+      required: false
+      secret: false
+    - auto: PREDEFINED
+      default: false
+      description: isGroupOwnerUpdate
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:969]: same here

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 7]
```diff
@@ -0,0 +1,1096 @@
...
+      description: Whether OATH two-factor authentication is enabled
+      isArray: false
+      name: oathtwofactor
+      predefined:
+      - 'true'
+      - 'false'
+      required: false
+      secret: false
+    - default: false
+      description: Password
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1004]: same here

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 8]
```diff
@@ -0,0 +1,1096 @@
...
+      required: false
+      secret: false
+    - default: false
+      description: RADIUS username
+      isArray: false
+      name: radiususername
+      required: false
+      secret: false
+    - default: false
+      description: timeOptionId (integer)
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1026]: same here

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 1 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:873]: @guykeller 
Replaced variable names

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 2 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:879]: done

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.py (hunk 3 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.py:500]: @guykeller 
**checkout/checkin**


Secret Server’s Check Out feature forces accountability by granting exclusive access to a single user with a One Time Password (OTP). OTP, also known as one-time pin or dynamic password, is a password that is valid for only one login session or transaction. OTPs avoid a number of risks associated with traditional, static passwords, while also provide the critical privileged access that certain users require to perform a specific task or function.

After ‘Check In,’ Secret Server automatically performs a random password change on the remote machine. No other user can access a secret while it’s checked out. This requirement establishes direct accountability for a machine that’s accessed during a specific time period.

Check Out ensures that Secret Server is used for all password changes and generates a complete audit trail of password use.

Advanced scripting capabilities allow Check Out to integrate with PowerShell hooks that can run before and after a secret is checked out. These hooks can be used for custom checks or actions to ensure environments are ready, or validate that specific conditions are met, before a user can access a secret.

**thycotic-secret-rpc-changepassword - Remote Passwoed Changing**
allows you to change the password on a remote machine with different policies

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 4 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1083]: returns a secret object according to API.
from API:
`

Secret summary

active: boolean

    Whether the secret is active
autoChangeEnabled: boolean

    Indicates whether or not this Secret an auto changing password
checkedOut: boolean

    Whether the secret is currently checked out
checkOutEnabled: boolean

    Indicates whether or not checkout is enabled for the Secret
createDate: string (date-time)

    When the Secret was created
daysUntilExpiration: integer (int32)

    How many days until this Secret expires
doubleLockEnabled: boolean

    Indicates whether or not DoubleLock is enabled for this password
extendedFields: object[]

    Any requested extended fields from a lookup request
    ISecretSummaryExtendedField
folderId: integer (int32)

    Containing folder ID
hidePassword: boolean

    Indicates if the launcher password is set to be hidden
id: integer (int32)

    Secret ID
inheritsPermissions: boolean

    Indicates if this Secret inherits permissions from its folder
isOutOfSync: boolean

    Out of sync indicates that a Password is setup for autochange and has failed its last password change attempt or has exceeded the maximum RPC attempts
isRestricted: boolean

    Whether the secret is restricted
lastAccessed: string (date-time)

    When the Secret was last viewed, only populated when scope is Recent
lastHeartBeatStatus: HeartbeatStatus

    Current status of heartbeat
lastPasswordChangeAttempt: string (date-time)

    Time of most recent password change attempt
name: string

    Secret name
outOfSyncReason: string

    Reason message if the secret is out of sync
requiresApproval: boolean

    Indicates if this Secret requires approval
requiresComment: boolean

    Indicates if this Secret requires comment
responseCodes: string[]

    ResponseCodes
    string
secretTemplateId: integer (int32)

    Secret template ID
secretTemplateName: string

    Name of secret template
siteId: integer (int32)

    SiteId`

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 5 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:959]: @guykeller 
All descriptions of commands and arguments are taken from the basic API documentation.
I myself have not used many and do not know their correct application.

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 8 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1026]: @guykeller 

documentation API
[API](https://updates.thycotic.net/secretserver/restapiguide/10.8/index.html)

Comment (guykeller) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1026]: I'm not sure why the API link is relevant, can you explain?
No matter what the API says, the description should at least include a short plain English description of the argument, that will explain what it means and how it should be used

Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1026]: I will request a more detailed description of the API from Thycotik, if they have one.

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 5 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:959]: IsApplicationAccount, true - if this is account usage for SDK

Comment (guykeller) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:959]: I don't understand what you mean by that, can you elaborate?

Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:959]: @guykeller 
I have requested a more detailed explanation of the API parameters.
In this case, a service account is used to connect to the SecretServer using the SDK

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 7 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1004]: Password for created user

**Comment by guykeller** [comment]
@AndyNikolaev Can you please explain your previous commit?

**Comment by AndyNikolaev** [comment]
@guykeller 
Good Day

I did not receive an answer on decoding the parameters of the REST API calls from Thycotic.
I can remove from the command parameters those assignment arguments that I do not understand.

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 9]
```diff
@@ -0,0 +1,1096 @@
...
+      type: String
+  - arguments:
+    - default: false
+      description: Secret ID
+      isArray: false
+      name: secret_id
+      required: true
+      secret: false
+    deprecated: false
+    description: Retrieved username from secret
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:54]: Let's rephrase this

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 10]
```diff
@@ -0,0 +1,1096 @@
...
+      type: String
+  - arguments:
+    - default: false
+      description: Search for secret using secret name.
+      isArray: false
+      name: search_name
+      required: true
+      secret: false
+    deprecated: false
+    description: Search ID secret by Name
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:69]: Let's rephrase this one as well. How about this one?
```suggestion
    description: Search a secret by its name.
```

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 11]
```diff
@@ -0,0 +1,1096 @@
...
+      type: String
+  - arguments:
+    - default: false
+      description: Search folder name
+      isArray: false
+      name: foldername
+      required: true
+      secret: false
+    deprecated: false
+    description: Search folder by  folder name
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:195]:
```suggestion
    description: Search a specific folder by name
```

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 4 above)
Comment (guykeller) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1083]: So there's an entire dictionary of values under this context path?

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 12]
```diff
@@ -0,0 +1,1096 @@
...
+      type: String
+  - arguments:
+    - default: false
+      description: Secret ID
+      isArray: false
+      name: secret_id
+      required: true
+      secret: false
+    - default: false
+      description: Password
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1072]:
```suggestion
      description: The new password
```

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml [hunk 13]
```diff
@@ -0,0 +1,1096 @@
...
+      description: Whether Duo two-factor authentication is enabled
+      isArray: false
+      name: duotwofactor
+      predefined:
+      - 'true'
+      - 'false'
+      required: false
+      secret: false
+    - default: false
+      description: E-mail
```
Comment (guykeller) [inline Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:926]: let's expand this description

**Review by guykeller — CHANGES_REQUESTED** [review]
@AndyNikolaev 
I've emphasized some notes, and I'm also reaching out to Thycotic for data about the following arguments:

1. AutoChangeEnabled
2.checkOutChangePasswordEnabled
3. DateOptionId
4. timeOptionId
5. radiususername
6. isGroupOwnerUpdate
7. IsApplicationAccount
8. GroupOwners

Please let me know if there are any more arguments that I've missed and you aren't familiar with, or you're not sure what they're for.
Thanks!

**Comment by guykeller** [comment]
> @AndyNikolaev Can you please explain your previous commit?

@AndyNikolaev ?

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 9 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:54]: @guykeller 
"extracting the password field from the required secret"
Is this wording appropriate?

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 10 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:69]: Done

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 11 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:195]: Done

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 12 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1072]: @guykeller 
A secret object is returned that contains all of the secret fields. But the object itself is one.

Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:1072]: Description for input parameter "password" - changed

File: Packs/Thycotic/Integrations/Thycotic/Thycotic.yml (hunk 13 above)
Comment (AndyNikolaev) [reply Packs/Thycotic/Integrations/Thycotic/Thycotic.yml:926]: User email address

**Comment by AndyNikolaev** [comment]
@guykeller 

1. AutoChangeEnabled
`Secret Server will automatically initiate a password change after a Secret expires or on a schedule.`
2.checkOutChangePasswordEnabled
`This flag means that when performing a check-in operation, the server will change the secret password to a new one.`
3. DateOptionId
4. timeOptionId
5. radiususername
6. isGroupOwnerUpdate
7. IsApplicationAccount
`This flag means that this account can only be used for access by tools that use the SDK developed by Tycotic for different languages.`

8. GroupOwners

**Comment by AndyNikolaev** [comment]
@guykeller 
see the changes made, please

**Comment by AndyNikolaev** [comment]
@guykeller 
Done

**Review by guykeller — APPROVED** [review]

**Pull request merged**